### ISSUE TITLE
architecture

### ISSUE THREAD

**Issue by 2junhyeok** [yugwangyeol/Mimic-Blocker#6]
- unet 형태로 변형
- noisy waveform을 직접 출력하도록 변경
- 학습 진행 중 (에폭 당 약 4시간)

**Comment by yugwangyeol** [comment]
1월 2일

- 성능 향상을 위한 Architecture 방향
- Dataset: VCTK 
- Conversion model: Free-VC
- Feature Extractor: WavLM

[Generator]

- [x] Unet wavform output+ $\lambda$ emb(0.2) 올리기 (→0.4/10) @2junhyeok 
- [x] Unet noise output @yugwangyeol 

[Discriminator]

- 모델 구조 : 기존 Generator 코드 (UNet X)
- [x] Conv 2d 사용 @yugwangyeol 
- [x] (dataset max로 padding or 3s cut) + last layer freeze @zziman

**Comment by yugwangyeol** [comment]
VCTK_Data로 학습하는 코드 gwang/VCTK_Train branch에 올려놨습니다.

- VCTK_Train_3sec은 3초로 잘라서 하는 파일입니다.
- VCTK_Train_org는 기존 코드에 데이터 셋 변환과 collate_fn만 바꾼 코드입니다.

참고하셔서 사용하시면 됩니다!
@zziman @2junhyeok

**Comment by yugwangyeol** [comment]
[Discriminator Conv2d]

```python
class Discriminator(nn.Module):
    def __init__(self):
        super(Discriminator, self).__init__()
        self.conv = nn.Sequential(
            # 입력: [batch_size, 1, n_mels, time]
            nn.Conv2d(1, 64, kernel_size=3, stride=1, padding=1),
            nn.LeakyReLU(0.2, inplace=True),
            nn.MaxPool2d(kernel_size=2, stride=2),
            
            nn.Conv2d(64, 128, kernel_size=3, stride=1, padding=1),
            nn.LeakyReLU(0.2, inplace=True),
            nn.MaxPool2d(kernel_size=2, stride=2),
            
            nn.Conv2d(128, 256, kernel_size=3, stride=1, padding=1),
            nn.LeakyReLU(0.2, inplace=True),
            nn.MaxPool2d(kernel_size=2, stride=2),
        )
        
        self.flatten = nn.Flatten()
        self.adaptive_pool = nn.AdaptiveAvgPool2d((4, 4))

    def forward(self, x):
        # x shape: [batch_size, 1, n_mels, time]
        x = x.view(x.size(0), 1, 80, -1)  # n_mels=80으로 reshape
        x = self.conv(x)
        x = self.adaptive_pool(x)
        x = self.flatten(x)
        
        # 동적으로 Linear 레이어의 입력 크기를 설정
        linear_input_size = x.shape[1]
        linear_layer = nn.Linear(linear_input_size, 1).to(x.device)
        x = linear_layer(x)
        
        return torch.sigmoid(x)
```

- 위와 같은 형태로 만들어 학습 진행
- 학습 진행 중 (에폭 당 약 1시간 40분)

**Comment by zziman** [comment]
[Discriminator]
-  dataset max padding + last layer freeze

### main.py
```
import torch
import torchaudio
from torch.utils.data import DataLoader
from transformers import WavLMModel
from model import Discriminator, Generator
from train import train_noise_encoder
import os

# 장치 설정
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# 모델 저장 경로 설정
save_dir = './checkpoints'
if not os.path.exists(save_dir):
    os.makedirs(save_dir)

# 하이퍼파라미터
batch_size = 2  # 배치 크기
num_epochs = 10
learning_rate = 0.001
lambda_emb = 0.2
lambda_wav = 0.8

# 데이터 로드
train_dataset = torchaudio.datasets.VCTK_092(root="./Data", download=True)

def get_dataset_max_length(dataset):
    """데이터셋의 최대 길이를 계산"""
    print("데이터셋 최대 길이 계산 중...")
    max_length = 0
    for waveform, *_ in dataset:
        max_length = max(max_length, waveform.shape[1])
    print(f"데이터셋 최대 길이: {max_length}")
    return max_length

def collate_fn(batch, max_length):
    """전체 데이터셋의 최대 길이로 패딩"""
    waveforms = []
    for waveform, *_ in batch:
        if waveform.shape[1] < max_length:
            padding = torch.zeros(1, max_length - waveform.shape[1])
            waveform = torch.cat([waveform, padding], dim=1)
        waveforms.append(waveform)
    return torch.stack(waveforms)

# 데이터셋의 최대 길이 계산 및 시퀀스 길이 설정
dataset_max_length = get_dataset_max_length(train_dataset)
sequence_length = dataset_max_length  # Discriminator에 필요한 sequence_length 정의

# collate_fn에 최대 길이 전달
train_loader = DataLoader(
    train_dataset, 
    batch_size=batch_size, 
    shuffle=True, 
    collate_fn=lambda b: collate_fn(b, sequence_length)
)

# 모델 초기화
wavlm = WavLMModel.from_pretrained("microsoft/wavlm-large").to(device)
generator = Generator().to(device)
discriminator = Discriminator(input_length=sequence_length).to(device)  # 수정: stride 및 AdaptiveAvgPool1d 사용

# 더미 forward pass 및 출력 크기 확인
dummy_input = torch.randn(1, 1, sequence_length).to(device)
conv_output = discriminator.conv(dummy_input)
print(f"Conv output shape: {conv_output.shape}")

# 옵티마이저 초기화
optimizer_G = torch.optim.Adam(generator.parameters(), lr=learning_rate)
optimizer_D = torch.optim.Adam(discriminator.parameters(), lr=learning_rate)

# 학습 실행
img_losses, emb_losses, gan_losses = train_noise_encoder(
    generator, discriminator, wavlm, train_loader, 
    optimizer_G, optimizer_D, num_epochs, batch_size, 
    device, lambda_wav, lambda_emb
)

# 학습 완료 메시지 출력
print("학습 완료!")

# 모델 저장
torch.save(generator.state_dict(), os.path.join(save_dir, 'generator.pth'))
torch.save(discriminator.state_dict(), os.path.join(save_dir, 'discriminator.pth'))
print(f"모델이 {save_dir}에 저장되었습니다.")
```

### model.py
```
import torch
import torch.nn as nn

class Generator(nn.Module):
    def __init__(self):
        super(Generator, self).__init__()
        self.model = nn.Sequential(
            nn.Conv1d(1, 64, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.Conv1d(64, 128, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.ConvTranspose1d(128, 64, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.ConvTranspose1d(64, 1, kernel_size=3, padding=1),
            nn.Tanh()
        )
        
    def forward(self, x):
        return self.model(x)

class Discriminator(nn.Module):
    def __init__(self, input_length):
        super(Discriminator, self).__init__()
        self.conv = nn.Sequential(
            nn.Conv1d(1, 64, kernel_size=5, stride=2, padding=2),  # stride=2로 시퀀스 길이 절반
            nn.LeakyReLU(0.2, inplace=True),
            nn.Conv1d(64, 128, kernel_size=5, stride=2, padding=2),  # stride=2로 시퀀스 길이 절반
            nn.LeakyReLU(0.2, inplace=True),
            nn.Conv1d(128, 256, kernel_size=5, stride=2, padding=2),  # stride=2로 시퀀스 길이 절반
            nn.LeakyReLU(0.2, inplace=True),
        )
        self.pool = nn.AdaptiveAvgPool1d(256)  # 고정된 크기로 줄이기
        self.flatten = nn.Flatten()

        # Conv 레이어 출력 크기 자동 계산
        dummy_input = torch.randn(1, 1, input_length)
        with torch.no_grad():
            conv_output = self.conv(dummy_input)
            pooled_output = self.pool(conv_output)
            flattened_size = pooled_output.numel()

        # Linear 레이어 생성
        self.linear_layer = nn.Linear(flattened_size, 1)

        # Freeze 설정 (가중치와 바이어스 고정)
        self.linear_layer.weight.requires_grad = False  # 가중치 고정
        self.linear_layer.bias.requires_grad = False  # 바이어스 고정

    def forward(self, x):
        x = self.conv(x)
        x = self.pool(x)
        x = self.flatten(x)
        x = self.linear_layer(x)
        return torch.sigmoid(x)
```

- 위와 같이 코드 수정하여 진행 (메모리 에러로 배치 사이즈 2로 줄임)
- dataset max: 794730
- 1 에폭에 약 9시간 걸려 데이터셋 분포 확인
![image](https://github.com/user-attachments/assets/22ff0381-0235-4166-8ba0-985bad4454e0)
- dataset max length를 이상치로 판단하여 데이터셋 3초로 진행

**Comment by zziman** [comment]
[Discriminator]
- dataset 3s + last layer freeze

### model.py
```
class Discriminator(nn.Module):
    def __init__(self):
        super(Discriminator, self).__init__()

        # Conv 레이어 정의
        self.conv = nn.Sequential(
            nn.Conv1d(1, 64, kernel_size=5, stride=2, padding=2),  # in_channels=1
            nn.LeakyReLU(0.2),
            nn.Conv1d(64, 128, kernel_size=5, stride=2, padding=2),
            nn.LeakyReLU(0.2),
            nn.Conv1d(128, 256, kernel_size=5, stride=2, padding=2),
            nn.LeakyReLU(0.2),
            nn.Conv1d(256, 512, kernel_size=5, stride=2, padding=2),
            nn.LeakyReLU(0.2)
        )

        self.fc = None  # fc 레이어를 처음에는 None으로 설정
        self.final_layer = nn.Linear(64, 1)

        # 마지막 레이어 고정
        self.final_layer.weight.requires_grad = False
        self.final_layer.bias.requires_grad = False

    def forward(self, x):
        batch_size = x.size(0)
        x = self.conv(x)

        # fc 레이어 입력 크기를 자동으로 설정
        if self.fc is None:
            flattened_size = x.view(batch_size, -1).size(1)
            self.fc = nn.Linear(flattened_size, 64).to(x.device)

        x = x.view(batch_size, -1)
        x = torch.relu(self.fc(x))
        x = self.final_layer(x)
        return torch.sigmoid(x)
```

- VCTK_Train_3sec 파일 가져와서 진행
- dataset.py except 구문에서 에러 발생하여 수정함
```
class VCTKDataset(Dataset):
    def __init__(self, root, sample_rate=16000, duration=3):
        #try:
        self.dataset = torchaudio.datasets.VCTK_092(root=root, download=True)
                
        self.sample_rate = sample_rate
        self.target_length = duration * sample_rate
```
- 학습 진행 중 (1 에폭 당 20분 정도 소요됨)
- wav loss가 0으로 떨어지는 문제 발생하여 가중치 값 바꿔보는 중

```
class Discriminator(nn.Module):
    def __init__(self):
        super(Discriminator, self).__init__()

        # Conv 레이어 정의
        self.conv = nn.Sequential(
            nn.Conv1d(1, 64, kernel_size=5, stride=2, padding=2),  # in_channels=1
            nn.LeakyReLU(0.2),
            nn.Conv1d(64, 128, kernel_size=5, stride=2, padding=2),
            nn.LeakyReLU(0.2),
            nn.Conv1d(128, 256, kernel_size=5, stride=2, padding=2),
            nn.LeakyReLU(0.2),
            nn.Conv1d(256, 512, kernel_size=5, stride=2, padding=2),
            nn.LeakyReLU(0.2)
        )

        # Fully connected 레이어 정의
        self.fc = nn.Linear(481280, 64)  # 고정된 입력 크기
        self.final_layer = nn.Linear(64, 1)

    def forward(self, x):
        batch_size = x.size(0)
        x = self.conv(x)

        # Flattened size 출력
        flattened_size = x.view(batch_size, -1).size(1)
        # print(f"Flattened size: {flattened_size}")

        # fc 레이어 사용
        x = x.view(batch_size, -1)
        x = torch.relu(self.fc(x))
        x = self.final_layer(x)
        return torch.sigmoid(x)
```
- 고정된 입력 크기로 실험 재진행
- 공격 실패

**Comment by yugwangyeol** [comment]
> [Discriminator Conv2d]
> 
> ```python
>     def forward(self, x):
>         # x shape: [batch_size, 1, n_mels, time]
>         x = x.view(x.size(0), 1, 80, -1)  # n_mels=80으로 reshape
>         x = self.conv(x)
>         x = self.adaptive_pool(x)
>         x = self.flatten(x)
>         
>         # 동적으로 Linear 레이어의 입력 크기를 설정
>         linear_input_size = x.shape[1]
>         linear_layer = nn.Linear(linear_input_size, 1).to(x.device)
>         x = linear_layer(x)
>         
>         return torch.sigmoid(x)
> ```

해당 코드에서  x = self.adaptive_pool(x) 부분이 잘못 들어가서 다음과 같이 수정했습니다.

```python
def forward(self, x):
        # x shape: [batch_size, 1, n_mels, time]
        x = x.view(x.size(0), 1, 80, -1)  # n_mels=80으로 reshape
        x = self.conv(x)
        x = self.flatten(x)
        
        # 동적으로 Linear 레이어의 입력 크기를 설정
        linear_input_size = x.shape[1]
        linear_layer = nn.Linear(linear_input_size, 1).to(x.device)
        x = linear_layer(x)
        
        return torch.sigmoid(x)
```

**Comment by yugwangyeol** [comment]
[Unet noise output]

```python
noisy_waveforms = torch.clamp(generator(waveforms), -1, 1)
```
이전 코드에서 다음과 같이 수정 후 학습 진행

```python
noise = generator(waveforms)  # [batch_size, 1, time]
noisy_waveforms = torch.clamp(waveforms + noise, -1, 1)
```

- 학습 진행 중 (에폭 당 약 4시간 30분)

**Comment by yugwangyeol** [comment]
**[Model 1]**

- Discriminator
    - wav + Conv1d
    - vctk dataset 동적
- Generator : Original
- Train
    - wavlm

**[Model 2]**

- Discriminator
    - mel + Conv2d
    - vctk dataset 동적
- Generator : Original
- Train
    - wavlm

**[Model 3]**

- Discriminator
    - Conv2d
    - vctk dataset 동적
- Generator : U-Net
- Train
    - wavlm
    

**[Model 4]**

- Discriminator
    - mel + Conv2d
    - vctk dataset 동적
- Generator : Original
- Train
    - wavlm
    - Intermediate Embedding

**Comment by yugwangyeol** [comment]
기존 Unet 모델 실패했습니다.
따라서 3번 모델 제외하고 4번 진행하겠습니다.

**Comment by yugwangyeol** [comment]
**[Model 4]**

- Discriminator : mel + Conv2d
- Data : vctk dataset 동적
- Generator : Original
- Train :  wavlm + Intermediate Embedding(12,17,23)

- 학습 진행 중 (에폭 당 약 2시간)

**Comment by yugwangyeol** [comment]
Model 4 성공

[Model 5]

Discriminator : mel + Conv2d

Data : vctk dataset 동적

Generator : Unet

Train : wavlm + Intermediate Embedding(12,17,23)

학습 진행 중 (에폭 당 약 2시간 30분 소요)

**Comment by yugwangyeol** [comment]
[Model 2]

Discriminator: mel + Conv2d
vctk dataset: 동적
Generator : Original
Train: wavlm

최종모델 선정
